library/src/blas1/rocblas_scal.hpp

```diff
@@ -0,0 +1,111 @@
+/* ************************************************************************
+ * Copyright 2016-2019 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+#include "handle.h"
+#include "rocblas.h"
+#include "utility.h"
+
+template <typename T, typename U, typename V>
+__global__ void scal_kernel(rocblas_int n,
+                            V           alpha_device_host,
+                            rocblas_int inca,
+                            U           xa,
+                            rocblas_int offsetx,
+                            rocblas_int incx,
+                            rocblas_int stridex)
+{
+    T*        x     = load_ptr_batch(xa, hipBlockIdx_y, offsetx, stridex);
+    auto      alpha = load_scalar(alpha_device_host, hipBlockIdx_y, inca);
+    ptrdiff_t tid   = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+
+    // bound
+    if(tid < n)
+        x[tid * incx] *= alpha;
+}
+
+template <rocblas_int NB, typename T, typename U, typename V>
+rocblas_status rocblas_scal_template(rocblas_handle handle,
+                                     rocblas_int    n,
+                                     const V*       alpha,
+                                     rocblas_int    inca,
+                                     U              x,
+                                     rocblas_int    offsetx,
+                                     rocblas_int    incx,
+                                     rocblas_int    stridex,
+                                     rocblas_int    batch_count)
+{
+    // Quick return if possible. Not Argument error
+    if(n <= 0 || incx <= 0 || batch_count <= 0)
+    {
+        if(handle->is_device_memory_size_query())
```

Comment:
Note that `*_template` functions are not supposed to allocate device memory or do argument error-checking. If they do that, then they cannot be called from other rocBLAS functions (besides their C wrappers and/or other wrappers).

`<name>_impl`: Argument error-checking, logging, and device memory allocation. Called from `extern "C"` wrappers.

`<name>_template`: No error-checking, logging, or device memory allocation. If device memory is needed, it is passed by pointer. Can be called from other rocBLAS routines, which are responsible for allocating device memory.

The separation of `<name>_impl` and `<name>_template` is to allow you to call `<name>_template` from other rocBLAS code without duplicating device memory allocation (which, for speed, is only allowed once per rocBLAS call), without duplicating log entries of routines the rocBLAS user didn't call directly, and to avoid the overhead of arugment-checking, which should only be done at the highest-level routine.

In general, `.hpp` files for implementing `rocBLAS` should be avoided except when the kernel is needed outside of the `.cpp` file, in which case, a declaration rather than definition is preferred. The `.cpp` file should instantiate all necessary template argument combinations, either implicitly as with the C wrappers, or explicitly in case they're needed outside of the `.cpp` file.